managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java

```diff
@@ -1033,25 +1041,78 @@ public void asyncReadEntriesWithSkipOrWait(int maxEntries, long maxSizeBytes, Re
             OpReadEntry op = OpReadEntry.create(this, readPosition, numberOfEntriesToRead, callback,
                     ctx, maxPosition, skipCondition);
 
-            if (!WAITING_READ_OP_UPDATER.compareAndSet(this, null, op)) {
-                op.recycle();
-                callback.readEntriesFailed(new ManagedLedgerException.ConcurrentWaitCallbackException(), ctx);
-                return;
-            }
+            synchronized (pendingReadOpMutex) {
+                if (!WAITING_READ_OP_UPDATER.compareAndSet(this, null, op)) {
+                    op.recycle();
+                    callback.readEntriesFailed(new ManagedLedgerException.ConcurrentWaitCallbackException(), ctx);
+                    return;
+                }
 
-            if (log.isDebugEnabled()) {
-                log.debug("[{}] [{}] Deferring retry of read at position {}", ledger.getName(), name, op.readPosition);
+                if (log.isDebugEnabled()) {
+                    log.debug("[{}] [{}] Deferring retry of read at position {}", ledger.getName(), name,
+                            op.readPosition);
+                }
+
+                // Check again for new entries after the configured time, then if still no entries are available
+                // register to be notified.
+                if (getConfig().getNewEntriesCheckDelayInMillis() > 0) {
+                    delayCheckForNewEntriesTask = new DelayCheckForNewEntriesTask(op, callback, ctx);
+                } else {
+                    // If there's no delay, check directly from the same thread
+                    checkForNewEntries(op, callback, ctx);
```

Comment:
is it possible that this results in a deadlock since this is called while holding `pendingReadOpMutex`?